### ISSUE TITLE
Init containers do not ensure ceph OSD has registered credentials in ceph

### ISSUE THREAD

**Issue by dkowsley** [rook/rook#3186]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

Something caused my cluster to lose OSD credentials for 36 of my 91 OSDs. The operator pod may have been deleted by someone running an improperly scoped command (edit: or, more likely, the node/rack it was in lost connectivity and it was rescheduled). When it came back online it fired off all of the workers to ensure the OSDs were all configured. I can only assume something in this phase cleared the credentials for the OSDs but may have broken at some point and not added credentials back in (I was running from master at the time). The OSD pods do not ensure their credentials are registered with ceph upon startup.

**Expected behavior:**

The init containers should ensure the OSD's credentials are registered with ceph.

**How to reproduce it (minimal and precise):**

Delete the credentials for the OSD in ceph. Restart the OSD. (Yes, it sounds crazy but something in master at the time did cause this)

**Environment**:
* OS (e.g. from /etc/os-release):
Ubuntu 16.04 LTS
* Kernel (e.g. `uname -a`):
4.4.0-142-generic
* Cloud provider or hardware configuration:
Specific details upon request.
* Rook version (use `rook version` inside of a Rook Pod):
v0.9.0-238.g4f97270
* Kubernetes version (use `kubectl version`):
1.13.3
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
baremetal deployed via kubespray
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):
Now? HEALTH_OK

**Comment by travisn** [comment]
It's not clear what the root cause of this issue is, but being able to recover automatically shouldn't be a stretch.
@alfredodeza Could `c-v --activate` verify that the auth exists for the osd each time the osd is started? Since c-v is creating the osd auth when the osd is provisioned, it seems natural that the osd could be healed if something happens to the auth.

**Comment by christianhuening** [comment]
@dkowsley I have the same issue! Yesterday we had network issues and after the dust settled, the rook operator had restarted and 2 OSDs are now crashlooping. Their auth keys indeed are missing now. 
My cluster's also healthy now with 70/72 OSDs up and in.

How did you resolve the situation? Simply removing the OSDs and have them readded by the operator?

**Comment by travisn** [comment]
@liewegas What would cause the OSD auth key to disappear? This has happened in several clusters and it's not clear what the cause is. 

To repair it, do we just add the auth key back in with the command mentioned in the [add osd docs](http://docs.ceph.com/docs/nautilus/rados/operations/add-or-rm-osds/#adding-an-osd-manual)? 
```
ceph auth add osd.{osd-num} osd 'allow *' mon 'allow rwx' -i /var/lib/ceph/osd/ceph-{osd-num}/keyring
```

**Comment by liewegas** [comment]
The key can get removed by ``ceph auth rm osd.NNN`` or (more likely?) by ``ceph osd purge osd.NNN``.  If it's the latter, you'll also see that the OSD is removed from the CRUSH map (i.e., missing from ``ceph osd crush tree``).

Adding the auth key back should be sufficient, because when the OSD comes up it will add itself back to the CRUSH map on its own.

**Comment by liewegas** [comment]
BTW the cluster's audit log should tell you whether/which of the auth rm or osd purge commands was run.  IIRC by default it's going into the mons' stderr log output prefixed by ``audit``.  Or it'll be in the mons' log dir as ``ceph.audit.log``.

**Comment by christianhuening** [comment]
@travisn where would I execute that command? I tried a few things but either cannot access the keyring (from toolbox pod) or have no connection to ceph (from osd pod on same node).
I think i did something like that in the past, but cannot remember it ;-) Can you give me a headsup?

**Comment by dkowsley** [comment]
@christianhuening @travisn my repair method was to modify the deployments for the OSDs to sleep instead of starting the OSD. I run a cluster that has strict security so the following step was more complicated than it would be for standard deployments. After the PODs were up again, I ran a script that went to each host to docker exec into the containers to mount the device and extract the credentials. With a single file with all of these credentials, I then went into the tools pod to readd all of the keys to ceph essentially with the command that @travisn posted. Once that was completed, I restored the deployments to their original form. I had to manually `in` the afflicted OSDs via the tools pod. Once that was done, the cluster then went into healing and came back fully online. There was no dataloss or corruption, only the initial moment of panic and some tedious work.

Thanks for your help with this in slack a month ago, @travisn.

**Comment by dkowsley** [comment]
@liewegas are you suggesting that something in rook ran either the `osd purge` or `osd auth rm`?

**Comment by liewegas** [comment]
> @liewegas are you suggesting that something in rook ran either the `osd purge` or `osd auth rm`?

No idea if it was rook or a human or something else... but i'm guessing *something* did one command or the other to get the cluster in this state.  Checking the cluster logs (before they rotate away!) would help confirm?

**Comment by travisn** [comment]
If rook decided to remove an osd it should be much more destructive than just removing the auth. And since the cluster CR wasn't updated, the operator wouldn't have hit this case. Rook doesn't remove the osd auth by itself either. Though to confirm it would be helpful to see the operator logs around the time this happens next.

**Comment by christianhuening** [comment]
@travisn I still have the logs from our logging system and was able to find the spot in time where it happened:

```
2019-05-28 08:45:53.948659 I | op-osd: processing removed osd rook-ceph-osd-11
2019-05-28 08:45:54.554468 I | exec: Running command: ceph osd crush reweight osd.11 0.0 --cluster=rook-ceph-hdd --conf=/var/lib/rook/rook-ceph-hdd/rook-ceph-hdd.config --keyring=/var/lib/rook/rook-ceph-hdd/client.admin.keyring --format json --out-file /tmp/453408104
2019-05-28 08:45:54.936688 I | exec: reweighted item id 11 name 'osd.11' to 0 in crush map
2019-05-28 08:45:55.643118 I | exec: marked out osd.11. 
(lot's of "osd.11 is still acting for pg3.22b)
2019-05-28 10:06:16.650042 I | op-k8sutil: Removed deployment rook-ceph-osd-11
2019-05-28 10:06:22.663073 I | exec: Running command: ceph osd crush rm osd.11 --cluster=rook-ceph-hdd --conf=/var/lib/rook/rook-ceph-hdd/rook-ceph-hdd.config --keyring=/var/lib/rook/rook-ceph-hdd/client.admin.keyring --format json --out-file /tmp/546873768
2019-05-28 10:06:22.662957 I | op-k8sutil: confirmed rook-ceph-osd-11 does not exist
2019-05-28 10:06:23.961233 I | exec: Running command: **ceph auth del osd.11** --cluster=rook-ceph-hdd --conf=/var/lib/rook/rook-ceph-hdd/rook-ceph-hdd.config --keyring=/var/lib/rook/rook-ceph-hdd/client.admin.keyring --format json --out-file /tmp/487969511
2019-05-28 10:06:23.961076 I | exec: removed item id 11 name 'osd.11' from crush map
2019-05-28 10:06:25.283347 I | exec: removed osd.11
2019-05-28 12:39:34.009109 I | op-osd: started deployment for osd 11 (dir=false, type=bluestore)
2019-05-28 12:42:03.579218 I | op-osd: deployment for osd 11 already exists. updating if needed
``` 

So apparently rook thought osd.11 (and 15) should've been removed from the cluster and they in fact ran `ceph auth del`. As I said we had network issues with heavy disconnection errors. But how could that possibly cause rook to think we don't want that osd anymore?

**Comment by leseb** [comment]
It seems that Rook determined nodes were removed from the cluster thus triggering osds deletion...
@christianhuening do you have the logs before `2019-05-28 08:45:53.948659 I | op-osd: processing removed osd rook-ceph-osd-11`?

Thanks!

**Comment by dkowsley** [comment]
Looking at the logs provided by @christianhuening, I have to say that those logs very much align with what I experienced. I will check the crush map to check the weights to see if it modified mine, as well. Thank you for your logs, Christian. Otherwise, I wouldn't have thought to check the crush weights. 

@liewegas, If this happens again, I'll make sure to get a log dump of all of the rook logs as well as investigate the cluster logs to see what happened. Unfortunately, by the time I discovered the issue in my cluster, the k8s api logs had already rotated away due to configuration error that kept the log history too short. Due to updates and reboots in between the incident and noticing it had even occurred, I don't have logs for rook pods. The cluster, at the time, was very much in development and we didn't have centralized logging in place while we were testing if rook would satisfy our needs (it does, by the way!).

@leseb raises a good theory: it did seem as if rook had determined that some nodes had left the cluster permanently. The OSDs that were broken were isolated to ~8 of the 21 nodes. However, every node in the cluster has the exact same age in k8s. Does rook try to auto-heal ceph if it thinks a node has been lost? If so, can we disable this feature?

**Comment by christianhuening** [comment]
@dkowsley you theory sounds solid and match what we experienced!

@leseb I am afraid the log rotation kicked in while I was on vacation :(

We're back to normal operation btw by re-adding the auth keys. Then the two rogue OSDs got back in.

**Comment by dkowsley** [comment]
Yep. Just confirmed our cluster topology. The issue occurred along a rack boundary. I had four devices per node and eight nodes in the second rack. All 36 of the OSDs that were broken were contained within that rack. UPS or switch failure for that rack explains the trigger. This should be relatively simple to replicate.

**Comment by dkowsley** [comment]
@travisn @leseb This seems like a pretty significant issue that shouldn't be marked wontfix.

**Comment by BlaineEXE** [comment]
@travisn is/was this fixed by your changes to remove all code in Rook relating to OSD removal?

**Comment by travisn** [comment]
Yes, if the root cause was that rook removed the OSDs, this would be fixed. But since there is some mystery still about the cause of the disappearing OSDs it's difficult to say. Maybe we should close this as not necessary and reopen if someone hits it again.

**Comment by BlaineEXE** [comment]
Let's do that. We believe the issue will be fixed by removing Rook's ability to delete OSDs from Ceph. Feel free to reopen if further information comes available that this is not fixed in Rook v1.3.
